content/posts/api-contracts-definitons.md

```diff
@@ -0,0 +1,126 @@
+
+Obviously, OpenAPI not being directly integrated into the implementation frameworks has a great disadvantage: It does not enforce (e.g., at compile time) that your implementation actually perfectly fulfills the specified contract. However, you can achieve a similar outcome by adding a check for your code's compliance to your automation pipeline, which prevents releases that diverge from their contract in an unwanted manner.
+
+### GraphQL
+
+[GraphQL](https://graphql.org/) calls itself "a query language for your API". The technology is about defining a schema which strongly types your endpoint methods and the objects they expect and return.
+
+A simple GraphQL schema might look as follows:
+
+```graphql
+type Query {
+  # Returns a pokemon
+  pokemon(id: ID!): Pokemon
+}
+
+type Pokemon {
+  id: ID!
+  name: String!
+}
+```
+
+It is not only much more concise than the above OpenAPI specification, but also has great advantages because it is part of the GraphQL specification. Almost every GraphQL endpoint exposes its schema automatically, which is a direct product of the endpoints it actually exposes. This allows clients to query the contract directly from the endpoint and therefore know that it is always up-to-date. Tests can be run against that exposed schema, which would detect breaking changes automatically and potentially prevent releasing such. These conventions of how the endpoint exposes its documentation allow us to use comprehensive client frameworks such as [`apollo-client`](https://www.apollographql.com/)
+
+With GraphQL, there are also frameworks that allow writing a schema first and generating the respective boilerplate code from it. A popular tool for doing so is [`gqlgen`](https://gqlgen.com/) in Go.
+
+### gRPC
+
+Another popular technology for declaring contracts is [gRPC](https://grpc.io/). It is based on [Protocol Buffers](https://github.com/protocolbuffers/protobuf), which is a way of specifying how to serialize structured data. The interface of a protocol buffer is defined in a file that might look like this:
+
+```protobuf
+service PokemonService {
+  // Returns a pokemon
+  rpc GetPokemon (GetPokemonRequest) returns (Pokemon) {}
+}
+
+message GetPokemonRequest {
+  string id = 1;
+}
+
+message Pokemon {
+  string id = 1;
+  string name = 2;
+}
+```
+
+One big difference of protocol buffers to the other mentioned technologies is that the exchanged data is not plain text, but rather a binary format. This makes them very performant but also harder to debug, which renders a crucial necessity for having a clearly defined schema and API. A compiler of such a Protocol Buffer file is built into the tool chain and lets you generate boilerplate code from the specification and enforce compliance with the defined contract.
```

Comment:
", which renders a crucial necessity for having a clearly defined schema and API" this sentence part is a bit odd. Suggestion: debug. This makes having a clearly defined shema and API crucial. (or "a necessity")